src/Client.php

```diff
@@ -248,11 +254,15 @@ public function __call(string $endpoint, array $params = [])
         // build the query string
         $query_string = $this->paramsToQueryString($params);
 
+        $uri = $endpoint . 'JSON?' . $query_string;
+
+        $this->lastUrlUsed = $this->url . '/' . $uri;
```

Comment:
Thank you for looking into this.

This function is not logging anything. It's just storing the last URL used in the request, independent of whether or not the request was successful. In fact, to me that would be an absolutely necessary base for any logging facility as debugging is only possible if also the failing URLs are saved.

To also store the status of the request (be it the status of the HTTP request, or be it the status of the data query) would be a different story and may be very useful for logging. However, you currently don't deal with the failures of the HTTP request at all. You just run the request and in the next line try to pars the JSON body. I assume that `$HttpClient->get()` would throw an exception if it fails? Otherwise it would be worth storing the HTTP status code **or** the `$HttpClient` itself so it could be further investigated by the calling software.

In respect to the question of whether call it an URI or an URL: In this project, you have been using URL so far (see `$this->url` for the [scheme and hostname]('https://secure.geonames.org')) which does make sense (as an URL must include both, the address of the resource, as well as the way of accessing it (http-over-ssl in this case). Whether or not the server is local or not has nothing to do whether an URL or URI is used (see https://danielmiessler.com/study/difference-between-uri-url/ for a nice pragmatic differentiation of the two). So in our case, both terms make sense equally, as the underlying functions use the term URL (in fact, it's called "cURL") while the wrapper around those functions, GuzzleHttp, use the term URI instead (which was why I chose `$uri` for the local variable name). But actually, it cannot be a local server, as you have hard-coded the "base_uri" to https://secure.geonames.org which would only ever be local if someone actually modifies their hosts file or DNS server.

Maybe instead of `lastUrlUsed()` we could use `lastUrlRequested` (or `lastUriRequested` for that matter). This would be more accurate, as that's what's done. It does not imply whether or not the request was successful or not. I personally would prefer the url-variant, rather than the uri-variant, as most people know, what an URL is but there seems to be more confusion around what a URI is. In this case, while it could be both, it **must** be an URL (as some specific URIs would not work) and hence "URL" is the more specific term. But that are just my two cents and I have no strong opinion about the naming as both variants are correct.

How would you like to proceed? What do you want me to change?

Reply:
Hey @martin-rueegg , I like the idea of using `lastUrlRequested`, we can start with this to complete this PR and I would be happy to see a more in-depth logging mechanism solution concept in a different PR.